ftwhttp/client_test.go

```diff
@@ -1,63 +1,112 @@
-		DestAddr: "example.com",
-		Port:     443,
-		Protocol: "https",
-	}
-
-	c, err := NewClient(NewClientConfig())
-	assert.NoError(t, err)
+func TestClientTestSuite(t *testing.T) {
+	suite.Run(t, new(clientTestSuite))
+}
 
-	err = c.NewConnection(*d)
-	assert.NoError(t, err, "This should not error")
-	assert.Equal(t, "https", c.Transport.protocol, "Error connecting to example.com using https")
+func (s *clientTestSuite) SetupTest() {
+	var err error
+	s.client, err = NewClient(NewClientConfig())
+	s.NoError(err)
+	s.Nil(s.client.Transport, "Transport not expected to be initialized yet")
 }
 
-func TestDoRequest(t *testing.T) {
-	d := &Destination{
-		DestAddr: "httpbin.org",
-		Port:     443,
-		Protocol: "https",
+func (s *clientTestSuite) TearDownTest() {
+	if s.ts != nil {
+		s.ts.Close()
 	}
+}
 
-	c, err := NewClient(NewClientConfig())
-	assert.NoError(t, err)
+func (s *clientTestSuite) httpHandler() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		resp := new(bytes.Buffer)
+		for key, value := range r.Header {
+			_, err := fmt.Fprintf(resp, "%s=%s,", key, value)
+			s.NoError(err)
+		}
+
+		_, err := w.Write(resp.Bytes())
+		s.NoError(err)
+	}
+}
 
-	req := generateBaseRequestForTesting()
+func (s *clientTestSuite) httpTestServer(secure bool) {
+	if secure {
+		s.ts = httptest.NewTLSServer(s.httpHandler())
+	} else {
+		s.ts = httptest.NewServer(s.httpHandler())
+	}
+}
 
-	err = c.NewConnection(*d)
-	assert.NoError(t, err, "This should not error")
+func (s *clientTestSuite) TestNewClient() {
+	s.NotNil(s.client.Jar, "Error creating Client")
+}
 
-	_, err = c.Do(*req)
+func (s *clientTestSuite) TestConnectDestinationHTTPS() {
+	s.httpTestServer(secureServer)
+	d, err := DestinationFromString(s.ts.URL)
+	s.NoError(err, "This should not error")
+	s.client.SetRootCAs(s.ts.Client().Transport.(*http.Transport).TLSClientConfig.RootCAs)
+	err = s.client.NewConnection(*d)
+	s.NoError(err, "This should not error")
+	s.Equal("https", s.client.Transport.protocol, "Error connecting to example.com using https")
+}
 
-	assert.Error(t, err, "This should return error")
+func (s *clientTestSuite) TestDoRequest() {
+	s.httpTestServer(secureServer)
+	d, err := DestinationFromString(s.ts.URL)
+	s.NoError(err, "This should not error")
+	s.client.SetRootCAs(s.ts.Client().Transport.(*http.Transport).TLSClientConfig.RootCAs)
+	req := generateBaseRequestForTesting()
+	err = s.client.NewConnection(*d)
+	s.NoError(err, "This should not error")
+
+	response, err := s.client.Do(*req)
+
+	// I'm getting consistently 400 Bad Request from the httpsTestServer's response,
+	// so I'm commenting this out for now.
+	// Example call used to test this:
+	// nc httpbin.org 80
+	// UNEXISTENT /bad/path HTTP/1.4
+	// Host: httpbin.org
+	// User-Agent: curl/7.88.1
+	// Accept: */*
+	//
+	// Will return HTTP/1.1 404 NOT FOUND from upstream.
+	// The same call to the httpsTestServer will return HTTP/1.1 400 Bad Request.
```

Comment:
I debugged this. The server expects two things for this to work properly:
1. The HTTP version must conform to spec. Currently you transfer `1.4` but it must be `HTTP/1.4`
2. The `Host` header must be present

Reply:
Genius. Thanks 💪

Reply:
Ok, still cannot make this fail. Somehow the outside query is making the request to fail, even if it is not failing. I mean, you just get a bad request response. I wonder how the previous assertion worked at all 🤔

Reply:
Your test server writes `http.StatusOK`, so the last line of the test should be (notice that I also fixed the order of paramaeters):
```
s.Equal(http.StatusOK, response.Parsed.StatusCode, "Error in calling website")
```

Your test server just doesn't care about the path. You could change it to care if you really wanted a 404.